manta-pay/src/crypto/hash/poseidon/parameter_generation/lfsr.rs

```diff
@@ -0,0 +1,195 @@
+        let mut head = 0;
+        // b0, b1 describes the field
+        append_bits(&mut init_sequence, &mut head, 2, 1u8);
+        // b2...=b5 describes s-box: we always use non-inverse s-box
+        append_bits(&mut init_sequence, &mut head, 4, 0b00000u8);
+        // b6...=b17 describes prime_num_bits
+        append_bits(&mut init_sequence, &mut head, 12, prime_num_bits);
+        // b18...=b29 describes width
+        append_bits(&mut init_sequence, &mut head, 12, width as u16);
+        // b30..=39 describes r_f (num_full_rounds)
+        append_bits(&mut init_sequence, &mut head, 10, r_f as u16);
+        // b40..=49 describes r_p (num_partial_rounds)
+        append_bits(&mut init_sequence, &mut head, 10, r_p as u16);
+        // b50..=79 describes the constant 1
+        append_bits(
+            &mut init_sequence,
+            &mut head,
+            30,
+            0b111111111111111111111111111111u128,
+        );
+        let mut res = GrainLFSR {
+            state: init_sequence,
+            prime_num_bits,
+            head,
+        };
+        res.init();
+        res
+    }
+
+    fn update(&mut self) -> bool {
+        let new_bit =
+            self.bit(62) ^ self.bit(51) ^ self.bit(38) ^ self.bit(23) ^ self.bit(13) ^ self.bit(0);
+        self.state[self.head] = new_bit;
+        self.head += 1;
+        self.head %= 80;
+        new_bit
+    }
+
+    fn init(&mut self) {
+        for _ in 0..160 {
+            self.update();
+        }
+    }
+
+    #[inline]
+    fn bit(&self, index: usize) -> bool {
+        self.state[(index + self.head) % 80]
+    }
+
+    pub fn get_bits(&mut self, num_bits: usize) -> Vec<bool> {
+        let mut res = Vec::new();
+
+        for _ in 0..num_bits {
+            // Obtains the first bit
+            let mut new_bit = self.update();
+
+            // Loop until the first bit is true
+            while !new_bit {
+                // Discards the second bit
+                let _ = self.update();
+                // Obtains another first bit
+                new_bit = self.update();
+            }
+
+            // Obtains the second bit
+            res.push(self.update());
+        }
+
+        res
+    }
+
+    pub fn get_field_elements_rejection_sampling<F>(&mut self, num_elems: usize) -> Vec<F>
+    where
+        F: ParamField,
+    {
+        assert_eq!(F::MODULUS_BITS as u64, self.prime_num_bits);
+
+        let mut res = Vec::new();
+        for _ in 0..num_elems {
+            // Performs rejection sampling
+            loop {
+                // Obtain n bits and make it most-significant-bit first
+                let mut bits = self.get_bits(self.prime_num_bits as usize);
+                bits.reverse();
+
+                // Constructs the number
+                if let Some(f) = F::try_from_bits_le(&bits) {
+                    res.push(f);
+                    break;
+                }
+            }
+        }
+
+        res
+    }

```

Comment:
Same comments as below and remove internal function docs that explain "what" is happening. Instead always try to answer a "why" question when adding internal function `NOTE` comments.